### PR TITLE
INT-524 update istio adapter to use the latest go-metrics package

### PR DESCRIPTION
**Pull request by baburajvelayudhan**
**Description**
go.mod updated to use the latest go-metrics package. 

1. Changes in wavefront.go and system_stats.go to use the new reporter. 

**Additional context**
{{ Add any other context about the change here, for example, Fixes #issue-number }}

### CODE REVIEW

File: wavefront/wavefront.go [hunk 1]
```diff
@@ -51,20 +55,70 @@ type (
...
+			FlushIntervalSeconds: int(cfg.FlushInterval),
+		}
+		sender, err := senders.NewDirectSender(directCfg)
+		if err != nil {
+			panic(err)
+		}
+
+		wa.reporter = wf.NewReporter(
+			sender,
+			application.New("istio", "istio"),
```
Comment (akodali18) [inline wavefront/wavefront.go]: Include hostTags (from line 67) when you create a reporter.  Host Tags should be sent along with all metrics reported by this reporter.
Also why is application.New("istio", "istio") needed ?

File: wavefront/wavefront.go [hunk 2]
```diff
@@ -51,20 +55,70 @@ type (
...
+		}
+
+		sender, err := senders.NewProxySender(proxyCfg)
+		if err != nil {
+			panic(err)
+		}
+
+		wa.reporter = wf.NewReporter(
+			sender,
+			application.New("istio", "istio"),
```
Comment (akodali18) [inline wavefront/wavefront.go]: same as above.

File: wavefront/wavefront.go [hunk 3]
```diff
@@ -181,8 +235,10 @@ func writeMetrics(cfg *config.Params, insts []*metric.InstanceMsg) {
 			if float64Val, err := translateToFloat64(value); err != nil {
 				log.Warnf("couldn't translate metric value: %s %v, err: %v", metricName, value, err)
 			} else {
-				gauge := wf.GetOrRegisterMetric(metricName, metrics.NewGaugeFloat64(), tags).(metrics.GaugeFloat64)
+				gauge := metrics.NewGaugeFloat64()
```
Comment (akodali18) [inline wavefront/wavefront.go]: Should we use https://github.com/wavefrontHQ/go-metrics-wavefront/blob/3fca929deab620bfd77b9749a5f0cf221c10e62b/reporting/reporter.go#L364 here ?

File: wavefront/wavefront.go [hunk 4]
```diff
@@ -191,7 +247,8 @@ func writeMetrics(cfg *config.Params, insts []*metric.InstanceMsg) {
 				log.Warnf("couldn't translate metric value: %s %v, err: %v", metricName, value, err)
 			} else {
 				deltaMetricName := wf.DeltaCounterName(metricName)
-				counter := wf.GetOrRegisterMetric(deltaMetricName, metrics.NewCounter(), tags).(metrics.Counter)
+				counter := metrics.NewCounter()
```
Comment (akodali18) [inline wavefront/wavefront.go]: same as above, Should we use https://github.com/wavefrontHQ/go-metrics-wavefront/blob/3fca929deab620bfd77b9749a5f0cf221c10e62b/reporting/reporter.go#L364 here?

File: wavefront/wavefront.go [hunk 5]
```diff
@@ -51,20 +55,70 @@ type (
...
 )
 
 // ensure that WavefrontAdapter implements the HandleMetricServiceServer interface.
 var _ metric.HandleMetricServiceServer = &WavefrontAdapter{}
 
 // createWavefrontReporter creates a reporter that periodically flushes metrics to Wavefront.
-func createWavefrontReporter(cfg *config.Params) {
+func (wa *WavefrontAdapter) createWavefrontReporter(cfg *config.Params) {
 	hostTags := map[string]string{"source": cfg.Source}
+
```
Comment (akodali18) [inline wavefront/wavefront.go]: Check if  WavefrontAdapter.reporterInitialized is false before proceeding to create a new reporter. I don't see that parameter being used elsewhere (please verify this).
Alternatively, can also check if WavefrontAdapter.reporter == nil , before creating a new one. (Also remove wa.reporterInitialized if its not being used)

File: wavefront/wavefront.go (hunk 4 above)
Comment (baburajvelayudhan) [reply wavefront/wavefront.go]: changes done as per comment.

File: wavefront/wavefront.go (hunk 3 above)
Comment (baburajvelayudhan) [reply wavefront/wavefront.go]: changes done as per comment.

File: wavefront/wavefront.go (hunk 5 above)
Comment (baburajvelayudhan) [reply wavefront/wavefront.go]: The bool flag is removed and the condition check is now on WavefrontAdapter.reporter == nil

File: wavefront/wavefront.go (hunk 1 above)
Comment (baburajvelayudhan) [reply wavefront/wavefront.go]: The mentioned variable, hostTags map has only the "Source" currently.  
In the writeMetrics call, the tags are sent along with the metrics in the GetOrRegisterMetric() call.  I believe you are referring to this. Please correct me if my understanding is incorrect.

Regarding, the hardcoded value "istio" for application and service, 

The application Tag is mandatory in the call to NewReporter() call and tags (app and service names) are not in the config.Params. 

Is there any other option to address this issue?

File: wavefront/wavefront.go (hunk 2 above)
Comment (baburajvelayudhan) [reply wavefront/wavefront.go]: -- same as above response on application Tags

File: wavefront/wavefront.go (hunk 1 above)
Comment (akodali18) [reply wavefront/wavefront.go:81]: Discussed this with Vikram, for now we can have both application and service as "wavefront-istio-adapter".  In future we can remove them - https://github.com/wavefrontHQ/go-metrics-wavefront/issues/22

File: wavefront/wavefront.go [hunk 6]
```diff
@@ -48,23 +52,72 @@ type (
...
 	} else if proxy := cfg.GetProxy(); proxy != nil {
 		addr, _ := net.ResolveTCPAddr("tcp", proxy.Address)
-		go wf.WavefrontProxy(metrics.DefaultRegistry, cfg.FlushInterval, hostTags, cfg.Prefix, addr)
+
+		// extract proxy ip and port from address
+		proxyInfo := strings.Split(addr.String(), ":")
+
+		// address must be in the form <proxyhost:port>
+		if len(proxyInfo) != 2 {
+			panic("Proxy Address and/or port number is mising")
```
Comment (akodali18) [inline wavefront/wavefront.go:94]: nit: "missing."

File: wavefront/wavefront.go (hunk 2 above)
Comment (akodali18) [reply wavefront/wavefront.go:116]: same change as above.

File: wavefront/wavefront.go (hunk 1 above)
Comment (akodali18) [reply wavefront/wavefront.go:81]: Regarding "hostTags",  In v0.9.0, they were a way to specify tags at the reporter level and specifying the source tag i guess, meaning any tag specified will be added to all the points flushed by this reporter. This seems to have changed with the latest "go-metrics" library.

With the latest version, i think you need to specify reporter level tags using a customMap inside application tags - 
https://github.com/wavefrontHQ/wavefront-sdk-go/blob/master/application/application.go#L11 
and source can be specified using - wf.Source(<source>)

In this case we don't have any reporter level tags, and you already specify the source using "wf.Source(cfg.Source)" . So basically, now, hostTags is being used only to "createSystemStatsReporter(<hostTags>)" at line 123. . Could you move the declaration closer to its usage.

Comment (baburajvelayudhan) [reply wavefront/wavefront.go]: comments incorporated

File: wavefront/wavefront.go (hunk 6 above)
Comment (baburajvelayudhan) [reply wavefront/wavefront.go]: comments incorporated

File: wavefront/wavefront.go (hunk 2 above)
Comment (baburajvelayudhan) [reply wavefront/wavefront.go]: comments incorporated

**Comment by baburajvelayudhan** [comment]
all comments incorporated

File: install/wavefront/templates/wavefront.yaml [hunk 7]
```diff
@@ -1,14 +1,14 @@
 # this config is created through command
-# mixgen adapter -c $GOPATH/src/istio.io/istio/mixer/adapter/wavefront/config/config.proto_descriptor -o $GOPATH/src/istio.io/istio/mixer/adapter/wavefront/config -s=false -n wavefront -t metric
+# mixgen adapter -c $GOPATHsrc/istio.io/istio/mixer/adapter/wavefront/config/config.proto_descriptor -o $GOPATHsrc/istio.io/istio/mixer/adapter/wavefront/config -s=false -n wavefront -t metric
 apiVersion: "config.istio.io/v1alpha2"
 kind: adapter
 metadata:
   name: wavefront
-  namespace: {{ .Values.namespaces.istio }}
```
Comment (akodali18) [inline install/wavefront/templates/wavefront.yaml:7]: is this change required ?

File: wavefront/config/wavefront.yaml [hunk 8]
```diff
@@ -1,5 +1,5 @@
 # this config is created through command
-# mixgen adapter -c $GOPATH/src/istio.io/istio/mixer/adapter/wavefront/config/config.proto_descriptor -o $GOPATH/src/istio.io/istio/mixer/adapter/wavefront/config -s=false -n wavefront -t metric
+# mixgen adapter -c $GOPATHsrc/istio.io/istio/mixer/adapter/wavefront/config/config.proto_descriptor -o $GOPATHsrc/istio.io/istio/mixer/adapter/wavefront/config -s=false -n wavefront -t metric
```
Comment (akodali18) [inline wavefront/config/wavefront.yaml:2]: is this change intended ?

File: install/wavefront/templates/wavefront.yaml [hunk 9]
```diff
@@ -1,14 +1,14 @@
...
+  namespace: istio-system
 spec:
   description: 
   session_based: false
   templates:
-  - metric.{{ .Values.namespaces.adapter }}
```
Comment (akodali18) [inline install/wavefront/templates/wavefront.yaml:12]: is this change intended ?

File: go.mod [hunk 10]
```diff
@@ -10,17 +10,18 @@ require (
...
 	github.com/spf13/cobra v0.0.0-20180821161202-6fd8e29b07d8 // indirect
 	github.com/spf13/pflag v1.0.1 // indirect
-	github.com/stretchr/testify v1.3.0 // indirect
-	github.com/wavefrontHQ/go-metrics-wavefront v0.9.0
+	github.com/wavefronthq/go-metrics-wavefront v1.0.2
+	github.com/wavefronthq/wavefront-sdk-go v0.9.5
 	go.uber.org/atomic v1.3.2 // indirect
 	go.uber.org/multierr v1.1.0 // indirect
 	go.uber.org/zap v1.9.1 // indirect
+	golang.org/x/tools v0.0.0-20200402223321-bcf690261a44 // indirect
```
Comment (akodali18) [inline go.mod:21]: we didn't need this dep earlier, where do we need this now

File: go.mod [hunk 11]
```diff
@@ -10,17 +10,18 @@ require (
...
 	google.golang.org/grpc v1.16.0
+	google.golang.org/protobuf v1.20.1 // indirect
```
Comment (akodali18) [inline go.mod:23]: we didn't need this dep earlier, where do we need this now

**Comment by baburajvelayudhan** [comment]
closing the pull request since https://github.com/vmware/wavefront-adapter-for-istio/pull/93 addresses all the suggested comments/ issues
